executor/src/native_executor.rs

```diff
@@ -15,20 +15,33 @@
 // along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
 
 use error::{Error, ErrorKind, Result};
-use native_runtime as runtime;
 use runtime_io;
 use state_machine::{Externalities, CodeExecutor};
 use wasm_executor::WasmExecutor;
 
 use std::panic::catch_unwind;
 
-pub struct NativeExecutor;
+/// Delegate for dispatching a CodeExecutor call to native code.
+pub trait NativeExecutionDispatch {
+	/// Get the wasm code that the native dispatch will be equivalent to.
+	fn native_equivalent() -> &'static [u8];
+	
+	/// Dispatch a method and input data to be executed natively. Returns `Some` result or `None`
+	/// if the `method` is unknown. Panics if there's an unrecoverable error.
+	fn dispatch(method: &str, data: &[u8]) -> Option<Vec<u8>>;
+}
+
+/// A generic `CodeExecutor` implementation that uses a delegate to determine wasm code equivalence
+/// and dispatch to native code when possible, falling back on `WasmExecutor` when not.
+pub struct NativeExecutor<D: NativeExecutionDispatch + Sync + Send> {
+	pub _dummy: ::std::marker::PhantomData<D>,
+}
 
-fn safe_call<F: ::std::panic::UnwindSafe + FnOnce() -> Vec<u8>>(f: F) -> Result<Vec<u8>> {
+fn safe_call<F: ::std::panic::UnwindSafe + FnOnce() -> Option<Vec<u8>>>(f: F) -> Result<Option<Vec<u8>>> {
 	catch_unwind(f).map_err(|_| ErrorKind::Runtime.into())
 }
 
-impl CodeExecutor for NativeExecutor {
+impl<D: NativeExecutionDispatch + Sync + Send> CodeExecutor for NativeExecutor<D> {
```

Comment:
does it compile if we change this line to `CodeExecutor for D`?

Reply:
no - neither `D` nor `CodeExecutor` are local.